internal/redactor/redactor.go

```diff
@@ -0,0 +1,425 @@
+	// where we've chosen to flush to inside a redacted range.
+
+	prevBufLen := len(r.buf)
+
+	// 1. Append b to the buffer.
+	r.buf = append(r.buf, b...)
+
+	// 2. Search through b to find instances of strings to redact. Store the
+	//    ranges of redactions in r.redact.
+	for n, c := range b {
+		bufidx := n + prevBufLen // where we are in the whole buffer
+
+		// In the middle of redacting?
+		for _, s := range r.states {
+			// Does the needle match on this byte?
+			if c != s.needle[s.matched] {
+				continue
+			}
+
+			// It matched!
+			s.matched++
+
+			// Have we fully matched this needle?
+			if s.matched < len(s.needle) {
+				// This state survives for another byte.
+				r.nextStates = append(r.nextStates, s)
+				continue
+			}
+
+			// Match complete; save range to redact.
+			r.redact = append(r.redact, subrange{
+				from: bufidx - len(s.needle) + 1,
+				to:   bufidx + 1,
+			})
+		}
+
+		// Start redacting something?
+		for _, s := range r.needlesByFirstByte[c] {
+			if len(s) == 1 {
+				// A pathological case; in practice we don't redact secrets
+				// smaller than RedactLengthMin.
+				r.redact = append(r.redact, subrange{
+					from: bufidx,
+					to:   bufidx + 1,
+				})
+				continue
+			}
+			r.nextStates = append(r.nextStates, state{
+				needle:  s,
+				matched: 1,
+			})
+		}
+
+		// r.nextStates contains the new set of states.
+		// Re-use the array underlying the old r.states for r.nextStates.
+		r.states, r.nextStates = r.nextStates, r.states[:0]
+	}
+
+	// 3. Merge overlapping redaction ranges.
+	// Because they were added from start to end, they are in order.
+	r.redact = mergeOverlaps(r.redact)
+
+	// 4. Write as much of the buffer as we can without spilling incomplete
+	//    matches.
+	flushTo := len(r.buf)
+	for _, s := range r.states {
+		if to := len(r.buf) - s.matched; to < flushTo {
+			flushTo = to
+		}
+	}
+	if err := r.flushTo(flushTo); err != nil {
+		// We "wrote" this much of b in this Write at the point of error.
+		return flushTo - prevBufLen, err
+	}
+
+	// We "wrote" all of b, so report len(b).
+	return len(b), nil
+}
+
+// Flush writes all buffered data to the destination. It assumes there is no
+// more data in the stream, and so any incomplete matches are non-matches.
+func (r *Redactor) Flush() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// We know all the complete matches in each Write, what we don't know
+	// are any incomplete matches. Since there is no more incoming data, any
+	// remaining states must all be incomplete.
+	r.states = r.states[:0]
+	return r.flushTo(len(r.buf))
+}
+
+// flush writes out the buffer up to an index. flushTo is an upper limit.
+func (r *Redactor) flushTo(flushTo int) error {
+	if flushTo == 0 || len(r.buf) == 0 {
+		return nil
+	}
+
+	bufidx := 0 // where we are up to in the buffer
+
+	// Stop when we're out of redactions, or the next one is after flushTo.
+	// Track the last range processed.
+	done := -1
+	for ri, rg := range r.redact {
```

Comment:
[whatever the question equivalent of a nit is]
`ri` is `redact index`, but what does `rg` stand for? is it a contraction of `range`?

Reply:
Given renaming `r.redact` to `r.completedMatches`, then `rg` seemed better as `match`.